### DISCUSSION TITLE
floating IPs

### DISCUSSION THREAD

**Discussion by michaelsstuff** [mysticaltech/terraform-hcloud-kube-hetzner#117] [Ideas]
Hi,

I do have the need to expose some services directly, without an LB. They are single pods and also are not HTTP. I figure that [hetzner cloud floating IP](https://docs.hetzner.com/cloud/floating-ips/faq/) should be perfect for this. 
I found this controller -> https://github.com/costela/hcloud-ip-floater - But it seems this does only manage existing ones, and would not create and attach a new floating ip to a service. 
Second I found ist this -> https://github.com/cbeneke/hcloud-fip-controller/ - But I could not get it running (very confusing docs...), but it also seems to only manage existing IPs.

Does anyone here have tried something like this? To fully automate the floating IP stuff.
I would love to see this integrated in this project

Thanks
best regards
Michael

**Comment by mysticaltech** [comment]
Why not use this @michaelsstuff? https://registry.terraform.io/providers/hetznercloud/hcloud/latest/docs/data-sources/floating_ip

Also why not do it manually, it would not change often, would it?

However, please know that the way we configure the LB with Traefik, we use proxy protocol... Meaning, it's as if the service itself was the LB. Meaning, there is actually no need to use a NodePort at all, just use the ingress controller, by opening a port of your choice via Traefik and its config (the LB never ever gets touched manually).

If you let me know more about your use case, I can probably help more.

**Reply by michaelsstuff** [reply]
The terraform module would only give information about the floating ip back to terraform. My problem is only in kuberntes, so maybe outside of this project?

Why not manully -> I do realy like fully automated deployments :-D
But there would still be the need to configure the floating ip to "follow" the node which runs the service or pod which could be done with the controller i posted.

Traefik would not work in my example, as an kubernetes ingress does not allow anything except http -> 

https://kubernetes.io/docs/concepts/services-networking/ingress/#what-is-ingress 
>An Ingress does not expose arbitrary ports or protocols. Exposing services other than HTTP and HTTPS to the internet typically uses a service of type [Service.Type=NodePort](https://kubernetes.io/docs/concepts/services-networking/service/#type-nodeport) or [Service.Type=LoadBalancer](https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer).

In my testcase, I wanted to run a teamskeak voice server. The service would expose UDP over 9987. There can only be one pod (no cluser or actual load balancing). So I created a deplyoment for 1 set with recreate and a service for it. But an ingress on top of it wont work, because UDP and arbitrary port. 

I think I could use a service with something like this:

```
status:
  loadBalancer:
    ingress:
    - ip: x.x.x.x
```

Adding the right labels would even create an actual hetzner LB. I just thought that creating and using the Floating IP would be better, as it matches the logic better and is slightly cheaper.

I hope this makes some sense? But if not, it is also not critical or anything, I am just curious how to go about a selfmade issue :smile:

**Reply by mysticaltech** [reply]
Ok, it will work with the default LB just add the udp entry point for that port to traefik_config.yaml... then create your ingress definition!

See https://doc.traefik.io/traefik/routing/entrypoints/

**Reply by mysticaltech** [reply]
Here's an example you can piggy back on for the traefik config of the entrypoint. https://traefik.io/blog/install-and-configure-traefik-with-helm/

**Reply by michaelsstuff** [reply]
:open_mouth: I did not know traefik could do that. Awseome, thank you!

**Reply by maggie44** [reply]
I’m trying to follow along with the setup that was used in the end. I too am exposing a UDP port for a pod. For example it listens on port 1234 and on a node with a Hetzner assigned ip 1.2.3.4, so I reach it on 1.2.3.4:1234. My load balancer is 5.6.7.8 and has a domain example.com pointed at it, but my understanding is Hetzner load balancers do not support UDP traffic. 

At the moment I manually set a subdomain that points at 1.2.3.4 and I change the subdomain ip whenever the ip changes of the server. 

Is there a better way?
